docs/record_reader.rst

```diff
@@ -0,0 +1,133 @@
+segment, implement the interface and plug it into the index engine - `SegmentCreationDriverImpl`.
+
+Initialize Record Reader
+------------------------
+
+To initialize a record reader, the data file and table schema should be provided (for Pinot segment record reader, only
+need to provide the index directory because schema can be derived from the segment). The output record will follow the
+table schema provided.
+
+For Avro/JSON/Pinot segment record reader, no extra configuration is required as column names and multi-values are
+embedded in the data file.
+
+For CSV/Thrift record reader, extra configuration might be provided to determine the column names and multi-values for
+the data.
+
+CSV Record Reader Config
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The CSV record reader config contains the following settings:
+
+- Header: the header for the CSV file (column names)
+- Column delimiter: delimiter for each column
+- Multi-value delimiter: delimiter for each value for a multi-valued column
+
+If no config provided, use the default setting:
+
+- Use the first row in the data file as the header
+- Use ',' as the column delimiter
+- Use ';' as the multi-value delimiter
+
+Thrift Record Reader Config
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The Thrift record reader config is mandatory. It contains the Thrift class name for the record reader to de-serialize
+the Thrift objects.
+
+Implement Your Own Record Reader
+--------------------------------
+
+You can implement your own record reader for file formats that is not supported natively. The following methods need to
+be implemented:
+
+.. code-block:: none
+
+  /**
+   * Return true if more records remain to be read.
+   */
+  boolean hasNext();
+
+  /**
+   * Get the next record.
+   */
+  GenericRow next()
+      throws IOException;
+
+  /**
+   * Get the next record. Re-use the given row if possible to reduce garbage.
+   * The passed in row should be returned by previous call to next().
+   */
+  GenericRow next(GenericRow reuse)
+      throws IOException;
+
+  /**
+   * Rewind the reader to start reading from the first record again.
+   */
+  void rewind()
+      throws IOException;
+
+  /**
+   * Get the Pinot schema.
+   */
+  Schema getSchema();
+
+
+Generic Row
+~~~~~~~~~~~
+
+Generic row is the record abstraction which the index engine can read and index with. It is a map from column name
+(String) to column value (Object). For multi-valued column, the value should be an object array (Object[]).
+
+Contracts for Record Reader
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There are several contracts for record readers that developers should follow when implementing their own record readers:
+
+- The output GenericRow should follow the table schema provided, in the sense that:
+
+  - All the columns in the schema should be preserved (if column does not exist in the original record, put default
+    value instead)
+  - Columns not in the schema should be escaped
```

Comment:
escaped? Is removed (not included) a better choice?

Reply:
Done